### PR TITLE
btrbk: 0.29.1 -> 0.31.3

### PR DESCRIPTION
**Pull request by woffs**
###### Motivation for this change
- new: lsbtr
- work around non-working --argv0 when wrapping perl or shell scripts
- add updateScript

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/backup/btrbk/default.nix [hunk 1]
```diff
@@ -36,11 +53,20 @@ stdenv.mkDerivation rec {
   '';
 
   preFixup = ''
-    wrapProgram $out/sbin/btrbk \
+    wrapProgram $out/bin/btrbk \
       --set PERL5LIB $PERL5LIB \
+      --run 'export program_name=$0' \
```
Comment (asymmetric) [inline pkgs/tools/backup/btrbk/default.nix:58]: Why was this necessary?

**Review by asymmetric — COMMENTED** [review]
Thanks!

File: pkgs/tools/backup/btrbk/default.nix (hunk 1 above)
Comment (woffs) [reply pkgs/tools/backup/btrbk/default.nix:58]: `exec -a` does set `$0` for the executed program. If the executed program itself is a (perl or shell) script, it apparently gets overwritten:
```
$ cat p
#!/bin/sh
echo "$0"
$ sh -c "exec -a dollarzero ./p"
/tmp/p
```
lsbtr is a symlink to btrbk, which itself is the wrapper and does `exec -a lsbtr $out/bin/.btrbk-wrapped`, and .btrbk-wrapped is a perl script and sets its $0 to .btrbk-wrapped and cannot find out if it is called as lsblk and behaves as btrbk even when called lsbtr.

My lines work around this and transfer the called executable's name in a separate variable.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 99303` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>btrbk</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 99303` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>btrbk</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
@woffs Can you solve the merge conflict?

**Comment by woffs** [comment]
> @woffs Can you solve the merge conflict?

done.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 99303` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>btrbk</li>
  </ul>
</details>

`gpl3` is a deprecated license, check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

**Comment by asymmetric** [comment]
It's [`gpl3Only`](https://github.com/digint/btrbk/blob/master/COPYING)

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 99303` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>btrbk</li>
  </ul>
</details>

File: pkgs/tools/backup/btrbk/default.nix [hunk 2]
```diff
@@ -36,11 +53,20 @@ stdenv.mkDerivation rec {
...
 
+  passthru.updateScript = genericUpdater {
+    inherit pname version;
+    versionLister = writeShellScript "btrbk-versionLister" ''
+      echo "# Versions for $1:" >> "$2"
+      ${curl}/bin/curl -s https://digint.ch/btrbk/download.html | ${perl}/bin/perl -lne 'print $1 if /<li>btrbk-([0-9.]*) /'
+    '';
+  };
+
   meta = with stdenv.lib; {
```
Comment (SuperSandro2000) [inline pkgs/tools/backup/btrbk/default.nix:70]:
```suggestion
  meta = with lib; {
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please fix the merge conflict.

**Review by asymmetric — COMMENTED** [review]
0.31.0 was released

File: pkgs/tools/backup/btrbk/default.nix [hunk 3]
```diff
@@ -36,15 +54,24 @@ stdenv.mkDerivation rec {
...
       --set PERL5LIB $PERL5LIB \
+      --run 'export program_name=$0' \
       --prefix PATH ':' "${lib.makeBinPath [ btrfs-progs bash mbuffer openssh ]}"
   '';
 
+  passthru.updateScript = genericUpdater {
+    inherit pname version;
+    versionLister = writeShellScript "btrbk-versionLister" ''
+      echo "# Versions for $1:" >> "$2"
+      ${curl}/bin/curl -s https://digint.ch/btrbk/download.html | ${perl}/bin/perl -lne 'print $1 if /<li>btrbk-([0-9.]*) /'
```
Comment (asymmetric) [inline pkgs/tools/backup/btrbk/default.nix:67]: How about we use [`list-git-tags`](https://github.com/asymmetric/nixpkgs/tree/134b5c9ac026465700678036250b63702eea1ff7/pkgs/common-updater/scripts/list-git-tags) instead? Seems less fragile than parsing HTML output.

See [here](https://github.com/asymmetric/nixpkgs/blob/134b5c9ac026465700678036250b63702eea1ff7/pkgs/applications/video/shotcut/default.nix#L82) for an example.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 99303` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>btrbk</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

btrbk:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unused --replace got detected:
```
substituteStream(): WARNING: pattern '/etc' doesn't match anything in file './doc/Makefile'
substituteStream(): WARNING: pattern '/bin/date' doesn't match anything in file 'contrib/cron/btrbk-mail'
substituteStream(): WARNING: pattern '/bin/echo' doesn't match anything in file 'contrib/cron/btrbk-mail'
```

Please check the offending substituteInPlace for typos or changes in source.
</details>

File: pkgs/tools/backup/btrbk/default.nix (hunk 3 above)
Comment (woffs) [reply pkgs/tools/backup/btrbk/default.nix:67]: I'm the pleased author of the example :-)
My thought is to parse download.html because it is expected to harmonize with the src attribute and because I'm always happy to have not a github reference.

Comment (asymmetric) [reply pkgs/tools/backup/btrbk/default.nix:67]: How about https://digint.ch/download/btrbk/releases/ ?

I appreciate the intent on relying (a tiny bit) less on GH, but I think parsing HTML is prone to breakage, and it would be hard to spot.

Comment (woffs) [reply pkgs/tools/backup/btrbk/default.nix:67]: Like
`curl -s https://digint.ch/download/btrbk/releases/ | perl -lne 'print $1 if /btrbk-([0-9.]*)\.tar/'`
?
